R/loadFromMothur.R

```diff
@@ -0,0 +1,232 @@
+
+    # Checks that colnames contain information, it is not NULL, and taxonomical information is present 
+    if ( !(length(colnames(data)) > 0) || is.null(colnames(data)) || is.null(data[[MOTHUR_TAX_COL]]) ){
+      stop("'taxonomy' does not include taxonomical information.",
+           call. = FALSE)
+    }
+    
+    # Removes additional characters between taxa
+    data [,MOTHUR_TAX_COL] <- gsub("[\"]", "", data [,MOTHUR_TAX_COL])
+    data [,MOTHUR_TAX_COL] <- gsub("[(1-100)]", "", data [,MOTHUR_TAX_COL])
+    # Splits taxa level into separate columns
+    tax <- tidyr::separate(data , MOTHUR_TAX_COL, into=c("Kingdom", "Phylum", "Order", "Class", "Family", "Genus"), sep=";", extra="merge")
+    # Removes ";" from the end of genus level names
+    tax$Genus <- gsub(";", "", tax$Genus)
+    
+    # Combines the extracted taxonomical data and unprocessed data and creates a data frame
+    rowData <- cbind(DataFrame(tax),
+                     DataFrame(data[, !(colnames(data) %in% MOTHUR_TAX_COL)]))
+    rowData <- tax
+    # If the data includes "OTU" column, that includes taxa
+    if( !is.null(rowData$OTU) ){
+      # Checks if the rownames are the same as in assay.
+      if( !identical(rowData$OTU, rownames(feature_tab)) ){
+        stop("taxa in 'taxonomyFile' does not match to taxa in 'sharedFile'.",
+             call. = FALSE)
+      }
+      # Adds rownames
+      rownames(rowData) <- rowData$OTU
+    }
+    return(rowData)
+}
+
+.read_mothur_sample_meta <- function(designFile, data_to_colData,...){
+  
+    if (.get_mothur_file_type(designFile) != "design") {
+      stop("The input '", designFile, "' must be in `design` format.",
+           call. = FALSE)
+    }
+  
+    # Reads the file
+    colData <- read.table(designFile, check.names=FALSE,
+                          header=TRUE, sep="\t", stringsAsFactors=FALSE)
+    
+    # Combines the extracted colData and data from the assay
+    colData <- cbind(colData, data_to_colData)
+    
+    # If the data includes 'group' column that includes the names of the samples
+    if( !is.null(colData$group) ){
+      # Checks if the sample names extracted from sample meta data ('group' column) 
+      # are the same as the ones that were extracted from assay ('Group' column)
+      if( !identical(colData$group, colData$Group) ){
+        stop("sample names in 'designFile' does not match to sample names in 'sharedFile'.",
+             call. = FALSE)
+      }
+      # Adds sample names to rownames of the data 
+      rownames(colData) <- colData$group
+    }
+    return(colData)
+  
+}
+
+#' extract file extension
+#' @noRd
+.get_mothur_file_type <- function(file) {
+    ex <- strsplit(basename(file), split = ".", fixed = TRUE)[[1]]
+    ex[length(ex)]
+}
+
+#' extract file extension
+#' @noRd
+.get_mothur_taxonomy_file_type <- function(file) {
+    ex <- strsplit(basename(file), split = ".", fixed = TRUE)[[1]]
+    paste(c(ex[length(ex)-1], ".", ex[length(ex)]), collapse= "")
+}
```

Comment:
These checks only work based on the name of the file. I think this is not really a solid solution, since problems can occur in to modes:
- wrong file name, right data
- right file name, wrong data

Mode 1 is not a severe problem, but mode 2 will produce unexpexted results. 

I would suggest the following:
- read in the data via `read.table`, because this is the common format (all files are tab-delimited files)
- check, if column names match certain assumptions. From what I can tell, this is actually the only requirement for correct processing of the data in following steps: A `shared` file requires the columns `"label","numOtus","Group"`, a taxonomy file either has the `Taxonomy` column or exactly `two` columns and the same number of rows than the shared file has data and a meta data file has column names and the same number of rows as the shared file has rows

Regarding the check implementation, it would be better to formulate the check differently. Instead of having a function return a type and do the check outside of the function, have a function like `.is_mothur_taxonomy_file` return a `boolean` result. That way the logic is contained in a single function and not separated into to two, the function and the check via `if`. This makes the code better maintainable and splits the logic for the individual files up. Each can be adjusted independently

Reply:
I modified those file type checks